pkg/service/presentation/storage/bolt.go

```diff
@@ -20,6 +20,63 @@ type BoltPresentationStorage struct {
 	db *storage.BoltDB
 }
 
+type opUpdater struct {
+	storage.UpdaterWithMap
+}
+
+func (u opUpdater) SetUpdatedResponse(first []byte) {
+	u.UpdaterWithMap.Values["response"] = first
+}
+
+func (u opUpdater) Validate(v []byte) error {
+	var op opstorage.StoredOperation
+	if err := json.Unmarshal(v, &op); err != nil {
+		return errors.Wrap(err, "unmarshalling operation")
+	}
+
+	if op.Done {
+		return errors.New("operation already marked as done")
+	}
+
+	return nil
+}
+
+var _ storage.ResponseSettingUpdater = (*opUpdater)(nil)
+
+func (b BoltPresentationStorage) UpdateSubmission(id string, approved bool, reason string, opID string) (StoredSubmission, opstorage.StoredOperation, error) {
+	m := map[string]any{
+		"status": StatusDenied,
+		"reason": reason,
+	}
+	if approved {
+		m["status"] = StatusApproved
+	}
+	submissionData, operationData, err := b.db.UpdateValueAndOperation(
+		submissionNamespace,
+		id,
+		storage.NewUpdater(m),
+		opstorage.Namespace,
+		opID,
+		opUpdater{
+			storage.NewUpdater(map[string]any{
+				"done": true,
+			}),
+		})
+	if err != nil {
+		return StoredSubmission{}, opstorage.StoredOperation{}, errors.Wrap(err, "updating value and operation")
```

Comment:
nit: consider returning pointers to avoid needing to instantiate empty objects

Reply:
I think there's a semantic difference between an empty object and a nil object (e.g. empty set vs no set at all)

I agree it makes the caller API a bit more annoying because of the ptr dereferencing issue. I really wish you could get around this in go with an Optional type. There is [a library](https://markphelps.me/posts/option-types-with-go-generate/) but it's fairly un-go-like.

one of my favorite parts of scala was [the option type](https://www.baeldung.com/scala/option-type) which I believe has made its way into java and kotlin too

Reply:
Definitely agree that it's very much not idiomatic go to use option type. Having used them in the past, I view `(*T, error)` in go as exactly the same. 

In Java world (I assume it's similar in Scala), there is a slight different between something like `Optional<int>` and `Optional<Integer>`, since in the latter, the object can be `null`. That's a similar problem as what we're discussing, I think.